### PR TITLE
Heap Sort - added return, tests, fixed readme

### LINKED ISSUE #10: heap_sort

### PR DESCRIPTION
**Pull request by delvor**
- [ ] PR's title includes the challenge's name and language
- [ ] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [ ] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [x] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [x] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [ ] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [ ] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [ ] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [x] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [x] Only one assertion
- [x] All tests pass
- [x] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/sorting/python/heap_sort/src/heap_sort.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    []
+    >>> heap_sort([3,15,5,8,19])
+    [3, 5, 8, 15, 19]
+    >>> merge_sort([1,12,9,20,6], False)
+    [20, 12, 9, 6, 1] """
+
+    assert isinstance(array_to_sort, (list, tuple)
+                      ), "array_to_sort must be list or tuple"
+
+    if len(array_to_sort) > 0:
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/src/heap_sort.py:30]: You can make 
if len(array_to_sort)==0:
   return []
instead

File: src/sorting/python/heap_sort/src/heap_sort.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        previous_element = array_to_sort[0]
+        for e in range(len(array_to_sort)):
+            if type(previous_element) != type(array_to_sort[e]):
+                raise TypeError("Array data types mismatch")
+            previous_element = array_to_sort[e]
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/src/heap_sort.py:36]: Good! 
previous_element = array_to_sort[e]  - is not needed (same will be to compare all to first)

but! What if you pass [2, 5, 3.2] - integers and floats? You can compare them, but their types mistmatch

File: src/sorting/python/heap_sort/src/heap_sort.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                swap_index = child_right
+        else:
+            if child_right < size and array[swap_index] > array[child_right]:
+                swap_index = child_right
+
+        # check if root is the largest, if not swap values in array and make recursive call
+        if swap_index != root:
+            swap(root, swap_index)
+            heapify(swap_index, size)
+
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/src/heap_sort.py:76]: Didn't understand it completely, but it looks like fine implemetation of heap algorythm, I read

File: src/sorting/python/heap_sort/src/heap_sort.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                      ), "array_to_sort must be list or tuple"
+
+    if len(array_to_sort) > 0:
+
+        previous_element = array_to_sort[0]
+        for e in range(len(array_to_sort)):
+            if type(previous_element) != type(array_to_sort[e]):
+                raise TypeError("Array data types mismatch")
+            previous_element = array_to_sort[e]
+
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/src/heap_sort.py:37]: def test_array_data_mixed_types_test(self):
        data = (10, 5, 8, 1.1)

        with self.assertRaises(TypeError):
            heap_sort(data)


This test pass, meaning (10, 5, 8, 1.1) will raise TypeError

**Comment by tvsirius** [comment]
Everything looks good, but type assertion in the beginging needs to be fixed, so function can work with list of mixed int and float items, and no docstrings in test suite

File: src/sorting/python/heap_sort/tests/test_heap_sort.py [hunk 5]
```diff
@@ -6,39 +7,49 @@
 file_dir = os.path.dirname(os.path.dirname(__file__))
 sys.path.append(file_dir + "/src")
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/tests/test_heap_sort.py:9]: in my VSCode it works only if i put 
from heap_sort import heap_sort
here
not in the begining,
and autoformatting moves it to begining

File: src/sorting/python/heap_sort/tests/test_heap_sort.py [hunk 6]
```diff
@@ -6,39 +7,49 @@
...
 
-from heap_sort import heapsort_v1
-
 
 class TestHeapSortV1(unittest.TestCase):
     def test_empty_array(self):
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/tests/test_heap_sort.py:12]: we need to add docstrigns for methods and classes and module as well

**Comment by delvor** [comment]
It meant to work only with single type arrays. There is no sense to mix int and float. They not comparable because of different types. Either float should be casted to int, but it will loose data beyond decimal or int should be casted to float which is anyway brings array to single data type.
About test they were provided. I just added additional. Not sure if we supposed to write any additional or leave it be how it is

**Comment by tvsirius** [comment]
Well in my issue i think i whould be better for more universal function, that can sort anything it can
But maybe you have the point, and we also don't know the context.
So then point it out in readme, and add docstrigns, and we merge

**Comment by delvor** [comment]
What docstring you mean? For unit test?

**Comment by tvsirius** [comment]
image
yes

File: src/sorting/python/heap_sort/tests/test_heap_sort.py (hunk 5 above)
Comment (tvsirius) [reply src/sorting/python/heap_sort/tests/test_heap_sort.py:9]: As I speak to Evan I see all this as not really serious things, and as this is a test task for learning 
I also don't find it nessesary to write the docstrings for tests, just i make them because of guides
Let's just move on )

Just we will ask about items assertions to know better

**Review by tvsirius — APPROVED** [review]
